src/components/promptBuilder/promptBuilder.tsx

```diff
@@ -0,0 +1,311 @@
+import Divider from '@mui/material/Divider'
+import Typography from '@mui/material/Typography'
+import { Box, useTheme } from '@mui/system'
+import React, { useEffect, useState } from 'react'
+
+import ElementRenderer from '../elementRenderer/elementRenderer'
+import Icon from '../icon/icon'
+import type { Message, PromptBuilderProps } from '../types'
+
+/**
+The `PromptBuilder` component is an interactive tool designed to help end users construct enhanced prompts through guided conversations with an agent. Users should be able to invoke the PromptBuilder via a button click or a text input command. Once activated, the agent guides the user step-by-step to create a detailed prompt, culminating in a "Generate" button for finalization or a "Next Question" option for further refinement.
+
+The expected output of the `PromptBuilder` should not be direct answers to the gathered input. Instead, when the "Generate" button is clicked, the component should produce refined prompt suggestion(s) based on the user's responses during the guided conversation. The `PromptBuilder` does not display the results directly; instead, developers can use the `onSubmit` function to handle the output and choose the most appropriate place to present it within their application.
+
+Try out the chat application example below to see how the `PromptBuilder` component can be used in your own application. Please note that the questions asked by agent are fixed as there is no backend involved. This is showcasing an example flow and use case.
+
+Notes for implementation:
+- The user will be allowed to start generating prompts when a message has been received with format of `"promptBuilder"` and data field of `isLastQuestion` set to `true`. For example:
+```
+const newMessage = {
+    ...otherRequiredFields,
+    format: 'promptBuilder',
+    data: {
+      isLastQuestion: true
+    }
+}
+```
+ */
+export default function PromptBuilder(props: PromptBuilderProps) {
+  const [isReadyToGenerate, setIsReadyToGenerate] = useState(false)
+  const [isLoading, setIsLoading] = useState(false)
+  const [isAttemptingToQuit, setIsAttemptingToQuit] = useState(false)
+  const [isGeneratingPrompts, setIsGeneratingPrompts] = useState(false)
+  const [messages, setMessages] = useState<Message[]>([])
+  const [lastInputMessage, setLastInputMessage] = useState<Message | null>(null)
+
+  const theme = useTheme()
+  const isMobileView = useMediaQuery(theme.breakpoints.down('md'))
+  const lastMessage = messages[messages.length - 1]
+
+  const inputCapturer = {
+    ...props.ws,
+    send: (message: Message) => {
+      setLastInputMessage(message)
+    },
+  }
+
+  useEffect(() => {
+    function handleIncomingMessage(event: MessageEvent) {
+      const receivedMessage = JSON.parse(event.data)
+      setMessages((prev) => [...prev, receivedMessage])
+    }
+
+    if (props.ws.onReceive) {
+      props.ws.onReceive(handleIncomingMessage)
+    }
+  }, [props.ws])
```

Comment:
Would it be better to move the function out of the useEffect?
```suggestion
  function handleIncomingMessage(event: MessageEvent) {
    const receivedMessage = JSON.parse(event.data)
    setMessages((prev) => [...prev, receivedMessage])
  }

  useEffect(() => {
    if (props.ws.onReceive) {
      props.ws.onReceive(handleIncomingMessage)
    }
  }, [props.ws])
```